vllm/model_executor/models/chatglm.py

```diff
@@ -0,0 +1,557 @@
+                                gather_output=False,
+                                perform_initialization=False,
+                                params_dtype=config.torch_dtype
+                            )
+        # rotary_dim=self.hidden_size_per_attention_head
+        self.atten = PagedAttention(config.num_attention_heads, 
+                                            self.hidden_size_per_attention_head, 
+                                            self.norm_factor)
+        # Output.
+        self.dense = RowParallelLinear(
+                        self.projection_size,
+                        config.hidden_size,
+                        bias=config.add_bias_linear,
+                        input_is_parallel=True,
+                        perform_initialization=False,
+                        params_dtype=config.torch_dtype
+                    )
+        # debug
+        # self.core_attention = CoreAttention(config, self.layer_number)
+
+    def forward(
+        self,
+        positions: torch.Tensor,
+        hidden_states: torch.Tensor,
+        kv_cache: KVCache,
+        input_metadata: InputMetadata,
+        cache_event: Optional[torch.cuda.Event],
+    ) -> torch.Tensor:
+        # hidden_states: [sq, b, h]
+
+        # =================================================
+        # Pre-allocate memory for key-values for inference.
+        # =================================================
+        # =====================
+        # Query, Key, and Value
+        # =====================
+
+        # Attention heads [sq, b, h] --> [sq, b, (np * 3 * hn)]
+        mixed_x_layer, _ = self.query_key_value(hidden_states)
+
+        if self.multi_query_attention:
+            (query_layer, key_layer, value_layer) = mixed_x_layer.split(
+                [
+                    self.num_attention_heads_per_partition * self.hidden_size_per_attention_head,
+                    self.num_multi_query_groups_per_partition * self.hidden_size_per_attention_head,
+                    self.num_multi_query_groups_per_partition * self.hidden_size_per_attention_head,
+                ],
+                dim=-1,
+            )
+            query_layer = query_layer.contiguous().view(
+                query_layer.size()[:-1] + (self.num_attention_heads_per_partition, self.hidden_size_per_attention_head)
+            )
+            key_layer = key_layer.contiguous().view(
+                key_layer.size()[:-1] + (self.num_multi_query_groups_per_partition, self.hidden_size_per_attention_head)
+            )
+            value_layer = value_layer.contiguous().view(
+                value_layer.size()[:-1]
+                + (self.num_multi_query_groups_per_partition, self.hidden_size_per_attention_head)
+            )
+        else:
+            new_tensor_shape = mixed_x_layer.size()[:-1] + \
+                               (self.num_attention_heads_per_partition,
+                                3 * self.hidden_size_per_attention_head)
```

Comment:
Nit: Please use paraphrases for line continuation.
```suggestion
            new_tensor_shape = (mixed_x_layer.size()[:-1] + 
                                (self.num_attention_heads_per_partition,
                                 3 * self.hidden_size_per_attention_head))
```